### DISCUSSION TITLE
Unexpected request method OPTIONS

### DISCUSSION THREAD

**Discussion by anthonyshew** [trpc/trpc#2095] [Q&A] [answered]
Hey, tRPC folks.

Loved your work here enough to take tRPC to production.

We're having a pretty major issue with the following, though.

```
{
  path: undefined,
  ctx: undefined,
  type: 'unknown',
  input: undefined,
  shape: {
    message: 'Unexpected request method OPTIONS',
    code: -32005,
    data: { code: 'METHOD_NOT_SUPPORTED', httpStatus: 405 }
  },
  error: TRPCError: Unexpected request method OPTIONS
      at Object.resolveHTTPResponse (/app/apps/web/.next/server/chunks/7239.js:439:13)
      at Object.nodeHTTPRequestHandler (/app/apps/web/.next/server/chunks/7239.js:231:44)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async /app/apps/web/.next/server/chunks/7239.js:77:5
      at async Object.apiResolver (/app/node_modules/next/dist/server/api-utils/node.js:185:9)
      at async NextNodeServer.runApi (/app/node_modules/next/dist/server/next-server.js:395:9)
      at async Object.fn (/app/node_modules/next/dist/server/base-server.js:496:37)
      at async Router.execute (/app/node_modules/next/dist/server/router.js:226:36)
      at async NextNodeServer.run (/app/node_modules/next/dist/server/base-server.js:606:29)
      at async NextNodeServer.handleRequest (/app/node_modules/next/dist/server/base-server.js:321:20) {
    originalError: undefined,
    code: 'METHOD_NOT_SUPPORTED',
    [cause]: undefined
  }
}
```

We followed most convention in the nextjs-prisma starter. [Our repo can be found here](https://github.com/kittrgg/kittr). The only thing that I can think of that we did that isn't in the docs is we had to change the `[trpc].ts` right for a `[[...trpc]].ts` route for the NextJS catch-all.

We actually can't reproduce this ourselves in all honesty and we're looking for answers. Was hoping we aren't the first to run into this. We're thinking this looks a lot like a preflight CORS request and this is somewhat confirmed by our logs.

Would appreciate any guidance if we've seen this coming from tRPC before.

**Comment by KATT** [comment, marked as answer]
Hey @anthonyshew, 

You don't seem to have set up CORS in https://github.com/kittrgg/kittr/blob/master/apps/web/src/pages/api/trpc/%5B%5B...trpc%5D%5D.api.ts, could be that?


Maybe tRPC should be sending a 200 on `OPTIONS`-requests? I'm honestly not sure what the right approach is here.

**Reply by KATT** [reply]
Maybe this can help you? https://github.com/vercel/next.js/blob/canary/examples/api-routes-cors/pages/api/cors.js

**Reply by anthonyshew** [reply]
I'm not quite sure how to work this into the tRPC createNextApiHandler. 🤔 
EDIT: Wait, of course I know how to do this.

**Reply by anthonyshew** [reply]
The problem turned out to be quite simple.

We aren't deploying to Vercel so we replaced the tRPC root URL with our own hardcoded one: https://github.com/kittrgg/kittr/blob/daf052a8566a73c8b83a722f90ca89a386738bd6/apps/web/src/pages/_app.page.tsx#L89

(Yes, this code is a mess, long story).

The reason we were getting CORS errors was that some of our users were on `https://www.kittr.gg` and some were on the apex domain. Our tRPC root URL was wired to the apex domain so the fetches coming from the `www` domain were freaking out. Now, we're redirecting everyone to our apex so we don't have this happen anymore.

Leaving this here just in case someone else happens to be as...questionably minded as me.

Thanks, KATT, for entertaining my issue. Love tRPC, really appreciate your work.

**Reply by KATT** [reply]
Thanks for your kind words and for sharing steps for people who might encounter the same issue! 😍

**Comment by anthonyshew** [comment]
Hey @KATT, thanks for the reply.

Are you describing CORS being set up at the Nextjs layer? I've never needed to do such a thing and would be surprised to need to now. I haven't had a CORS error in all my time using Next. 🤔  I'd also argue that the request is making it to the tRPC layer - so it's getting through the Nextjs Lambda-ish endpoint's layer well enough?

However, I also would be surprised to learn that tRPC would have some odd CORS constraint. Really why this one is perplexing.

I would think that a 200 on `OPTIONS` would satisfy CORS so that the subsequent actual data fetching request would be fulfilled but can't say for sure. It's still an intermittent error for us that we can't identify any patterns for. Some of our users are completely unaffected while others aren't.

A 200 for `OPTIONS`, of course, may mean doing something to your library that you're not wanting to which I also respect. And, seeing as you have plenty of people using this lib and it doesn't appear you've seen this before, I'd imagine I've done something bizarre and it's my problem.

But it _only_ happens on our TRPC router. That's the only consistency that we can identify.

**Reply by anthonyshew** [reply]
![image](https://user-images.githubusercontent.com/35677084/177396923-bac0d07b-582f-491f-ba1e-0b6a047aa9e8.png)

Furthering the idea that this is indeed CORS being strange. No path supplied, not a method that is in the spec, etc.

**Comment by Ankcorn** [comment]
I just ran into this on AWS Lambda using http apis!

I worked around it by changing
```
"ANY /trpc/{path+}": "functions/trpc.handler",
```

to this
```
"GET /trpc/{path+}": "functions/trpc.handler",
"POST /trpc/{path+}": "functions/trpc.handler"
```

In my sst.dev configuration

**Reply by KATT** [reply]
cc @anthonyshew

**Reply by anthonyshew** [reply]
Hm, interesting. I'm using Next so any of that stuff gets pre-configured for me. I don't have access to any of that.

**Reply by ctrlplusb** [reply]
Thanks for saving me this pain! Exact same issue. Luckily I am using SST too. Copy and paste. 😂

**Reply by dan-kimani** [reply]
This saved me an hour of pain

**Reply by unwrap-nicolas** [reply]
For me it solved using this Ankcorn solution and allowing all headers in cors:

```Typescript
  let cors: boolean | ApiCorsProps | undefined = {
    allowMethods: ['ANY'],
    allowHeaders: ['*'], //I changed it
    allowOrigins: ['http://localhost:4000']
  }
```

I'm using sst too

**Reply by jamesopstad** [reply]
This solution isn't ideal because it results in deploying two lambda functions rather than one. I've worked around responding to the OPTIONS request using `ANY /{proxy+}` by adding the following to `responseMeta`:

```ts
responseMeta(opts) {
  return opts.type === 'unknown' ? { status: 200 } : {};
}
```

I still get `TRPCError: Unexpected request method OPTIONS` in the logs though.

It does seem like having tRPC handle OPTIONS requests by default would be the best option. Are there any reasons not to do this?

**Reply by iota-pi** [reply]
This feels like poor design on the part of AWS to me; why is there no flag in the CORS configuration to allow it to take preference over proxy integrations?

However, I think it would also be good to have much clearer instructions in the TRPC lambda adapter docs page about CORS

**Comment by blacky-i** [comment]
I have also encountered with this error when was trying queries from Swagger. I will leave here for somebody else who will encountered this in NextJS with TRPC.

Before code changes in pages/api/trpc/[trpc].ts:
```typescript
import { createNextApiHandler } from "@trpc/server/adapters/next"

import { createContext } from "@/server/trpc/context"
import { appRouter } from "@/server/trpc/router"
// export API handler
export default createNextApiHandler({
  router: appRouter,
  createContext,
  onError:
    process.env.NODE_ENV === "development"
      ? ({ path, error }) => {
          console.error(`tRPC failed on ${path}: ${error}`)
        }
      : undefined,
})

```

After code changes:
```typescript
import { createNextApiHandler } from "@trpc/server/adapters/next"

import { createContext } from "@/server/trpc/context"
import { appRouter } from "@/server/trpc/router"
// export API handler

import type {
  NextApiRequest,
  NextApiResponse,
} from 'next/types';

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse,
) {
  // Run the middleware

  if (req.method === 'OPTIONS') {
    res.writeHead(200);
    return res.end();
  }


  const runTrpc = createNextApiHandler({
    router: appRouter,
    createContext,
    batching: {
      enabled: true
    },
    onError:
      process.env.NODE_ENV === "development"
        ? ({ path, error }) => {
          console.error(`tRPC failed on ${path}: ${error}`)
        }
        : undefined,
  })
  return runTrpc(req, res);
}
```
